core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,174 @@
+package ca.jrvs.apps.grep;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+
+  private String regex;
+  private String rootDir;
+  private String outputFile;
+  private Logger jgrepLogger;
+
+  public JavaGrepImp() {
+    jgrepLogger = LoggerFactory.getLogger(JavaGrepImp.class);
+  }
+
+  /**
+   * Top level search workflow.
+   *
+   * @throws IOException when file operations fail
+   */
+  @Override
+  public void process() throws IOException {
+    List<String> lines = new ArrayList<>();
+
+    for (File file : listFiles(rootDir)) {
+      for (String line : readLines(file)) {
+        if (containsPattern(line)) {
+          lines.add(file.getName() + ": " + line);
+        }
+      }
+    }
+    writeToFile(lines);
+  }
+
+  /**
+   * Traverse a given directory and return all files.
+   *
+   * @param rootDir input directory
+   * @return files under the rootDir
+   */
+  @Override
+  public List<File> listFiles(String rootDir) {
+    List<File> fileList = new ArrayList<>();
+    File root = new File(rootDir);
+    File[] filesToCheck;
+    if (root.isDirectory()) {
+      filesToCheck = root.listFiles();
+      if (filesToCheck != null) {
```

Comment:
Some redundancies - can do for(filesTocheck : root.listFiles) <-- if root.listFiles is NULL, should not execute the loop

Reply:
My understanding is that if root.listFiles returns null the foreach will hit a null pointer and die.
From the Java docs on File.listFiles:
`The array will be empty if the directory is empty. Returns null if this abstract pathname does not denote a directory, or if an I/O error occurs.`